fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -664,6 +684,210 @@ public void exec() throws Exception {
         }
     }
 
+    /**
+     * Compute the topological order of the fragment tree.
+     *
+     * @return multiple fragment groups. Each group should be delivered sequentially,
+     * and fragments in a group can be delivered concurrently.
+     */
+    private List<List<PlanFragment>> computeTopologicalOrderFragments() {
+        Queue<PlanFragment> queue = Lists.newLinkedList();
+        Map<PlanFragment, Integer> inDegrees = Maps.newHashMap();
+
+        PlanFragment root = fragments.get(0);
+        inDegrees.put(root, 0);
+        queue.add(root);
+        // Compute in-degree of each fragment.
+        while (!queue.isEmpty()) {
+            PlanFragment fragment = queue.poll();
+            for (PlanFragment child : fragment.getChildren()) {
+                Integer v = inDegrees.get(child);
+                if (v != null) {
+                    inDegrees.put(child, v + 1);
+                } else {
+                    inDegrees.put(child, 1);
+                    queue.add(child);
+                }
+            }
+        }
+
+        if (fragments.size() != inDegrees.size()) {
+            for (PlanFragment fragment : fragments) {
```

Comment:
can this situation happen ?

Reply:
When there are some bugs, it will happen. This log is used to help us find these bugs, and I have found two places where fragments forgets to link to the fragment tree.

Reply:
which cases?